### PR TITLE
Refine HybridConnectionListener.ControlConnection reconnect logic

### PR DESCRIPTION
**Pull request by dlstucki**
Anywhere something in HCL.ControlConnection needs a webSocket it should use ensureConnectTask then get the ClientWebSocket from that `CompletableFuture<ClientWebSocket>`. Under normal circumstances when it's already connected that task will be complete and will simply give the current webSocket. If the ControlConnection is already in the process of reconnecting and a parallel call is made which needs the webSocket too that parallel call will wait on the same connectAsyncTask which will give the new ClientWebSocket when completed.

### CODE REVIEW

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java [hunk 1]
```diff
@@ -538,26 +536,22 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
...
+			return connectTask.thenApply((webSocket) -> {
 					this.tokenRenewer.setOnTokenRenewed((token) -> this.onTokenRenewed(token));
-					this.receivePumpAsync();
-					success = true;
-				} finally {
-					if (!success) {
-						this.closeOrAbortWebSocketAsync(connectTask, new CloseReason(CloseCodes.UNEXPECTED_CONDITION,
-								"closing web socket connection because something went wrong trying to connect."));			
+					this.receivePumpAsync(); // This starts pumping but doesn't wait
+					return (Void)null;
```
Comment (bainian12345) [inline src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: >(Void) [](start = 12, length = 6)

Is this cast needed?

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java [hunk 2]
```diff
@@ -538,26 +536,22 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
...
+				}).whenComplete(($void, err) -> {
+					if (err != null) {
+						// catch (Throwable err) {
+						RelayLogger.throwingException(err, this.listener);
+						CloseReason closeReason = new CloseReason(CloseCodes.UNEXPECTED_CONDITION,
+								"closing web socket connection because something went wrong trying to connect.");
```
Comment (bainian12345) [inline src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: >web socket [](start = 17, length = 10)

Do you think saying "control web socket" would be more specific?

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 1 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: It wouldn't compile without either this cast, specifying `.<Void>.thenApply` or putting it into a local variable of type `CompletableFuture<Void>`.

---
In reply to: [258629544](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258629544) [](ancestors = 258629544)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 2 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: Sure, I don't think I actually touched the value of this String.

---
In reply to: [258630907](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258630907) [](ancestors = 258630907)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 1 above)
Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: That's an indication of maybe .thenApply isn't the best fit here. Will .thenAccept work?

---
In reply to: [258631053](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258631053) [](ancestors = 258631053,258629544)

Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: Ooooooh. Awesome info for me to learn! I'll try it out.

---
In reply to: [258632412](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258632412) [](ancestors = 258632412,258631053,258629544)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java [hunk 3]
```diff
@@ -676,34 +679,42 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
...
-					}
-				}).thenRun(() -> this.onOnline());
+				ClientWebSocket webSocket = new ClientWebSocket(this.listener.trackingContext, EXECUTOR);
+				return delayTask.thenCompose(($void) -> {
+					return webSocket.connectAsync(websocketUri, timeout, config).thenApply(($void2) -> webSocket)
+						.whenComplete((ws, err) -> {
+							if (err == null) {
+								this.onOnline();
+							}
+						});
```
Comment (bainian12345) [inline src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: I don't think this .whenComplete block is necessary and the "thenApply" should be enough. If there is an exception from websocket connecting, the "thenApply" wouldn't be run anyways.

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java [hunk 4]
```diff
@@ -676,34 +679,42 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
...
-						return this.webSocket.connectAsync(websocketUri, timeout, config);
-					}
-				}).thenRun(() -> this.onOnline());
+				ClientWebSocket webSocket = new ClientWebSocket(this.listener.trackingContext, EXECUTOR);
+				return delayTask.thenCompose(($void) -> {
+					return webSocket.connectAsync(websocketUri, timeout, config).thenApply(($void2) -> webSocket)
+						.whenComplete((ws, err) -> {
+							if (err == null) {
+								this.onOnline();
+							}
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: >} [](start = 7, length = 1)

should I put an explicit throw (with trace) in here?
```
.whenComplete((ws, err) -> {
       if (err == null) {
           this.onOnline();
!>   } else {
!>       throw RelayLogger.throwingException(err);
!>   }
```

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 3 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: That sounds good.

---
In reply to: [258635262](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258635262) [](ancestors = 258635262)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 4 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: This changed to a thenApply so this comment is out of date.

---
In reply to: [258635516](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258635516) [](ancestors = 258635516)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 1 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: Fixed in iteration 3 (iteration 2 was a merge with master).

---
In reply to: [258633157](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258633157) [](ancestors = 258633157,258632412,258631053,258629544)

File: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java (hunk 3 above)
Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java]: Fixed in iteration 3 (iteration 2 was a merge with master).

---
In reply to: [258635798](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258635798) [](ancestors = 258635798,258635262)

**Comment by bainian12345** [comment]
> 		CompletableFuture<?>[] closeTasks;

Just curious why was this declaration explicitly separated from the assignment below?

---
Refers to: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java:281 in c9248cd. [](commit_id = c9248cdcca25268d40ea90740a8ffec295638d04, deletion_comment = False)

**Comment by dlstucki** [comment]
> 		CompletableFuture<?>[] closeTasks;

Because the collection is grabbed inside the sync block, but used later outside the sync block.

---
In reply to: [465719611](https://github.com/Azure/azure-relay-java/pull/17#issuecomment-465719611) [](ancestors = 465719611)

---
Refers to: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java:281 in c9248cd. [](commit_id = c9248cdcca25268d40ea90740a8ffec295638d04, deletion_comment = False)

**Comment by bainian12345** [comment]
> 		CompletableFuture<?>[] closeTasks;

Oh, I see

---
In reply to: [465720122](https://github.com/Azure/azure-relay-java/pull/17#issuecomment-465720122) [](ancestors = 465720122,465719611)

---
Refers to: src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java:281 in f693cb4. [](commit_id = f693cb46300fd93f7eabedf0a85af9a5692039b0, deletion_comment = False)

**Pull request merged**